### ISSUE TITLE
TimeoutExceptions and disconnects at Cacheable<IUserMessage, ulong>.GetOrDownloadAsync()

### ISSUE THREAD

**Issue by TheBlue-1** [RogueException/Discord.Net#1742]
When i call `Cacheable<IUserMessage, ulong>.GetOrDownloadAsync()` i often get a TimoutException like this one:
```
1/6/2021 1:31:50 AM: System.TimeoutException: The operation has timed out.
   at Discord.Net.Queue.RequestBucket.EnterAsync(Int32 id, RestRequest request)
   at Discord.Net.Queue.RequestBucket.SendAsync(RestRequest request)
   at Discord.Net.Queue.RequestQueue.SendAsync(RestRequest request)
   at Discord.API.DiscordRestApiClient.SendInternalAsync(String method, String endpoint, RestRequest request)
   at Discord.API.DiscordRestApiClient.SendAsync[TResponse](String method, String endpoint, String bucketId, ClientBucketType clientBucket, RequestOptions options)
   at Discord.API.DiscordRestApiClient.GetChannelMessageAsync(UInt64 channelId, UInt64 messageId, RequestOptions options)
   at Discord.Rest.ChannelHelper.GetMessageAsync(IMessageChannel channel, BaseDiscordClient client, UInt64 id, RequestOptions options)
   at Discord.WebSocket.SocketTextChannel.GetMessageAsync(UInt64 id, RequestOptions options)
   at Discord.WebSocket.SocketTextChannel.Discord.IMessageChannel.GetMessageAsync(UInt64 id, CacheMode mode, RequestOptions options)
   at Discord.WebSocket.DiscordSocketClient.<>c__DisplayClass116_2.<<ProcessMessageAsync>b__3>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Discord.Cacheable`2.DownloadAsync()
   at Discord.Cacheable`2.GetOrDownloadAsync()
   ```
   
   also there are many disconnects which often happen after this Exception, but not exclusivly.
   The Timeout exception only occurs when calling this method.
   
   I don't think there is a problem with internet because im hosting on gcloud.
   
   
   maybe still the same issue as here #1336 
   i'm using latest nuget version (2.2.0)
   
   at the same call I also get an request exception sometimes:
```
1/6/2021 11:30:42 PM: System.Net.Http.HttpRequestException: An error occurred while sending the request.
 ---> System.IO.IOException: The response ended prematurely.
   at System.Net.Http.HttpConnection.FillAsync()
   at System.Net.Http.HttpConnection.ReadNextResponseHeaderLineAsync(Boolean foldedHeadersAllowed)
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithNtConnectionAuthAsync(HttpConnection connection, HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.DecompressionHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Discord.Net.Rest.DefaultRestClient.SendInternalAsync(HttpRequestMessage request, CancellationToken cancelToken, Boolean headerOnly)
   at Discord.Net.Rest.DefaultRestClient.SendAsync(String method, String endpoint, CancellationToken cancelToken, Boolean headerOnly, String reason)
   at Discord.Net.Queue.RestRequest.SendAsync()
   at Discord.Net.Queue.RequestBucket.SendAsync(RestRequest request)
   at Discord.Net.Queue.RequestQueue.SendAsync(RestRequest request)
   at Discord.API.DiscordRestApiClient.SendInternalAsync(String method, String endpoint, RestRequest request)
   at Discord.API.DiscordRestApiClient.SendAsync[TResponse](String method, String endpoint, String bucketId, ClientBucketType clientBucket, RequestOptions options)
   at Discord.API.DiscordRestApiClient.GetChannelMessageAsync(UInt64 channelId, UInt64 messageId, RequestOptions options)
   at Discord.Rest.ChannelHelper.GetMessageAsync(IMessageChannel channel, BaseDiscordClient client, UInt64 id, RequestOptions options)
   at Discord.WebSocket.SocketTextChannel.GetMessageAsync(UInt64 id, RequestOptions options)
   at Discord.WebSocket.SocketTextChannel.Discord.IMessageChannel.GetMessageAsync(UInt64 id, CacheMode mode, RequestOptions options)
   at Discord.WebSocket.DiscordSocketClient.<>c__DisplayClass116_2.<<ProcessMessageAsync>b__3>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Discord.Cacheable`2.DownloadAsync()
   at Discord.Cacheable`2.GetOrDownloadAsync()
```   
   
   
   All other requests are working fine.

**Comment by quinchs** [comment]
What event are you calling this in? and could you share your code for it?

**Comment by TheBlue-1** [comment]
I call it in an async function after the reaction added event. 
I'm not sure if the bug still exists because I have a try catch around the call since I found it.

```cs
{
_client.ReactionAdded += ClientReactionAdded;
}
private static Task ClientReactionAdded(Cacheable<IUserMessage, ulong> arg1, ISocketMessageChannel arg2, SocketReaction arg3) {
            new Func<Task>(async () => {
                if (arg3.UserId == Program.Client.BotUserId
                    || arg3.User.Value?.IsBot == true
                    || arg3.Message.IsSpecified && arg3.Message.Value.Author.Id != Program.Client.BotUserId
                    || arg1.HasValue && arg1.Value.Author.Id != Program.Client.BotUserId) {
                    return;
                }
                IUserMessage message;
                try {
                    message = await arg1.GetOrDownloadAsync();
                } catch (Exception) {
                    return;
                }

                if (message?.Author?.Id == Program.Client.BotUserId) {
                    await Honor.Instance.HonorUserWithReaction(message, arg3);
                }
            }).RunAsyncSave();

            return FakeTask;
        }
```

I am not rerurning my task because its execution could be too long and trigger a timeoutexception.

**Comment by quinchs** [comment]
Try just making a new task like so
```cs
_ = Task.Run(async () => code)
```
and returning a `Task.CompletedTask`

**Comment by TheBlue-1** [comment]
this is nearly what i do, FakeTask is a completed task and RunAsyncSave is running the task like you do but with error handling

**Comment by quinchs** [comment]
Does version 2.4.0 fix this?

**Comment by TheBlue-1** [comment]
it seems to be fixed in 2.4.0
i tested it over the last month
